pkg/ingester/stream_rate_calculator.go

```diff
@@ -58,30 +63,57 @@ func (c *StreamRateCalculator) updateLoop() {
 }
 
 func (c *StreamRateCalculator) updateRates() {
+	rates := make([]logproto.StreamRate, 0, defaultStripeSize)
+
 	for i := 0; i < c.size; i++ {
 		c.locks[i].Lock()
 		c.rates[i] = c.samples[i]
-		c.samples[i] = 0
+		c.samples[i] = logproto.StreamRate{}
+
+		sr := c.rates[i]
+		if sr.StreamHash != 0 {
+			rates = append(rates, logproto.StreamRate{
+				StreamHash:        sr.StreamHash,
+				StreamHashNoShard: sr.StreamHashNoShard,
+				Rate:              sr.Rate,
+			})
+		}
 		c.locks[i].Unlock()
 	}
-}
 
-func (c *StreamRateCalculator) RateFor(streamHash uint64) int64 {
-	i := streamHash & uint64(c.size-1)
+	c.rateLock.Lock()
+	defer c.rateLock.Unlock()
+
+	c.allRates = rates
+}
 
-	c.locks[i].RLock()
-	defer c.locks[i].RUnlock()
+func (c *StreamRateCalculator) Rates() []*logproto.StreamRate {
+	c.rateLock.RLock()
+	defer c.rateLock.RUnlock()
+
+	rates := make([]*logproto.StreamRate, 0, len(c.allRates))
+	for _, r := range c.allRates {
```

Comment:
hmm can't you just use `copy(src, dst)` here? if so, do you know if it is more performatic than iterating over etc?

Reply:
We can't `copy...` here because we're changing the  type from `logproto.StreamRate` to `*logproto.StreamRate`. I'm doing it this way so no pointers are ever shared among instances (I kept making memory leaks)

Reply:
oops, thanks! :smile: